starport/services/network/join.go

```diff
@@ -0,0 +1,167 @@
+package network
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	launchtypes "github.com/tendermint/spn/x/launch/types"
+	"github.com/tendermint/starport/starport/pkg/cosmosaddress"
+	"github.com/tendermint/starport/starport/pkg/cosmosutil"
+	"github.com/tendermint/starport/starport/pkg/events"
+)
+
+func (b *Builder) Join(
+	ctx context.Context,
```

Comment:
Why Join func is a method of Builder and not Blockchain struct's? We have a Publish method under Blockchain, and it makes sense to have Join under Blockchain. Any functionality related to creating a network for a chain should be under Blockchain.

As we move Join under Blockchain, we don't need to pass `launchID` and `chainHome` to this func anymore, Blockchain should already know it when it's initialized with the SourceLaunchID  option through the Builder.Blockchain().

Reply:
I think it's because we initially specified `Join` as being "stateless". `init` generate the gentx and `join` allows to send the requests without interacting with the environment

Initializing the `Blockchain` fetch the source and the chain launch information. If we introduce an option for `no-check` this is not a necessary step

Reply:
How about only fetching the source code when Blockchain.Publish() and Blockchain.Init() called and not when calling Blockchain.Join()?

Reply:
The thing I don't like is that we start to make the `Blockchain` structure very stateful. For some logic flow, we use it with an initialize `Source` but an uninitialized `LaunchID`, for some others we have an initialized `launchID` but an uninitialized `Source`.

This seems to become an over-engineered structure to me.

**However, from this current codebase, we define the verification logic in the same method as sending the request to SPN, so I would be more in favor to integrate in `Blockchain` for now and download the source anyway.**

In my mind, the methods that solely interacts with SPN and that are agnostic to the chain setup should be separated so they can be called without initializing the `Blockchain` structure.
In the example of `join`. `Blockchain` could provide methods to verify the requests (and that requires the chain to be set up) and the methods to send them are separated, so if communication with SPN must be done with interacting with the blockchain itself (in the case of no verification) to can be performed without initializing the blockchain structure

Reply:
I agree with @lubtd. The `Blockchain` struct handles too much information. We don't need to use it in the join command only to send messages to the SPN, and we can avoid cloning the repo into a temp folder every time we run each command. But if we need to run the [`cmd.ShowNodeID(ctx)` command](https://github.com/tendermint/starport/pull/1774#pullrequestreview-814644940), we need the repo folder and the binary, so we need to use the blockchain struct with a chain repo path.

Some options:

- Clone the repo only into the `init` command into the chain home directory, reuse it for the chain `join` command, and move the chain struct to the `Builder` instead of the `Blockchain`.
- Use the `Blockchain` structure as is and clone every time we run the command. It can be slow but is the easiest implementation
- Find another way to get the node ID, we can save into some config into the home folder with the `init` command and reuse it from the `join` command

Reply:
Why don't we change Blockchain name with Source and only init it when we need to deal with the source code or appd binary? Maybe the name is a bit misleading after all. 

Then, we move the code that interacts with SPN to Builder struct.

This also means that we need to move Blockchain.Publish to Builder.Publish as an example. Builder.Publish can init Source if needs the source code or appd and other Builder methods can also do the same.

Whatever decision we make, let's merge this PR first as it's I think. Then let's create another PR to implement our decision on the overall structure.

Reply:
Great, nice idea! I will use the blockchain structure, cloning the repo again, for we can test and merge the PR, and we can improve the architecture of everything in another one, for this one don't increase too much